### ISSUE TITLE
[feature request] Device Client File upload APIs don't include IoT Hub status codes in thrown exceptions

### ISSUE THREAD

**Issue by sanjay-90210** [Azure/azure-iot-sdk-java#1182]
**Is your feature request related to a problem? Please describe.**

While using DeviceClient.getFileUploadSasUri, the real cause of the issue is hidden and a generic IOException with the message "Sas URI response message had no payload" is thrown. This hides the real issue that caused the empty payload and so programmatically addressing the exception is a lot of guess work or near impossible.

The above exception can be thrown in the following scenarios that I am running into:
1. A storage container is not associated at the iot hub end for file uploads. ([link](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-file-upload#associate-an-azure-storage-account-with-iot-hub))

     - This gets mapped to HTTP 400 - BAD_FORMAT

2. The number of outstanding SAS URIs may be more than that is allowed by the quota and limits ([link](https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/azure-subscription-service-limits#iot-hub-limits))

**File upload operations** | 83.33 file upload initiations/sec/unit (5,000/min/unit) (for S3),  1.67 file upload initiations/sec/unit (100/min/unit) (for S1 and S2).  10,000 SAS URIs can be out for an Azure Storage account at one time. **10 SAS URIs/device can be out at one time.**

       - This gets mapped to HTTP 403 - TOO_MANY_DEVICES
 
3. The device may no longer exist just before the SAS URI is requested. The device may be deleted or disabled.

      - This gets mapped to HTTP 404 - HUB_OR_DEVICE_ID_NOT_FOUND

In all the above cases, the only exception that I get is a vanilla IOException that says "Sas URI response message had no payload" and so cannot take any situation specific action.

Even though the IotHubStatusCode that reports these specific cases reach all the way till the ResponseMessage in the code below, it is ignored and a plain IOException is thrown from the function.

```
    public FileUploadSasUriResponse getFileUploadSasUri(FileUploadSasUriRequest request) throws IOException
    {
        IotHubTransportMessage message = new IotHubTransportMessage(request.toJson());
        message.setIotHubMethod(IotHubMethod.POST);

        ResponseMessage responseMessage;
        httpsTransportManager.open();
        responseMessage = httpsTransportManager.getFileUploadSasUri(message);
        httpsTransportManager.close();

        if (responseMessage.getBytes() == null || responseMessage.getBytes().length == 0)
        {
            throw new IOException("Sas URI response message had no payload");
        }

        return new FileUploadSasUriResponse(new String(responseMessage.getBytes(), DEFAULT_IOTHUB_MESSAGE_CHARSET));
    }
```

**Describe the solution you'd like**

If the IotHubStatusCode that is there in the response is propagated along with the IOException, it will be easy to identify the real cause and take the appropriate actions. The easiest fix without writing a new exception class would be to set the cause of the IOException being thrown as in the code below.

```
    public FileUploadSasUriResponse getFileUploadSasUri(FileUploadSasUriRequest request) throws IOException
    {
        IotHubTransportMessage message = new IotHubTransportMessage(request.toJson());
        message.setIotHubMethod(IotHubMethod.POST);

        ResponseMessage responseMessage;
        httpsTransportManager.open();
        responseMessage = httpsTransportManager.getFileUploadSasUri(message);
        httpsTransportManager.close();

        if (responseMessage.getBytes() == null || responseMessage.getBytes().length == 0)
        {
            String errorStr = "Sas URI response message had no payload";
            throw new IOException(errorStr, IotHubStatusCode.getConnectionStatusException(
                responseMessage.getStatus(), errorStr));
        }

        return new FileUploadSasUriResponse(new String(responseMessage.getBytes(), DEFAULT_IOTHUB_MESSAGE_CHARSET));
    }

```

**Describe alternatives you've considered**

Throw a new Exception that extends IOException and include the IotHubStatusCode in it.

**Additional context**

Adding a nested cause will allow the nested exception to retrieved and the real cause extracted. Some stack traces are given below.

```
Exception in thread "main" java.io.IOException: Sas URI response message had no payload
	at com.microsoft.azure.sdk.iot.device.fileupload.FileUploadTask.getFileUploadSasUri(FileUploadTask.java:188)
	at com.microsoft.azure.sdk.iot.device.DeviceClient.getFileUploadSasUri(DeviceClient.java:583)
	at com.kalkitech.test.azure.dfd.Main.main(Main.java:36)
Caused by: com.microsoft.azure.sdk.iot.device.exceptions.HubOrDeviceIdNotFoundException: Sas URI response message had no payload
	at com.microsoft.azure.sdk.iot.device.IotHubStatusCode.getConnectionStatusException(IotHubStatusCode.java:49)
	... 3 more
```

```
Exception in thread "main" java.io.IOException: Sas URI response message had no payload
	at com.microsoft.azure.sdk.iot.device.fileupload.FileUploadTask.getFileUploadSasUri(FileUploadTask.java:188)
	at com.microsoft.azure.sdk.iot.device.DeviceClient.getFileUploadSasUri(DeviceClient.java:583)
	at com.kalkitech.test.azure.dfd.Main.main(Main.java:36)
Caused by: com.microsoft.azure.sdk.iot.device.exceptions.TooManyDevicesException: Sas URI response message had no payload
	at com.microsoft.azure.sdk.iot.device.IotHubStatusCode.getConnectionStatusException(IotHubStatusCode.java:46)
	... 3 more

```

```
Exception in thread "main" java.io.IOException: Sas URI response message had no payload
	at com.microsoft.azure.sdk.iot.device.fileupload.FileUploadTask.getFileUploadSasUri(FileUploadTask.java:188)
	at com.microsoft.azure.sdk.iot.device.DeviceClient.getFileUploadSasUri(DeviceClient.java:583)
	at com.kalkitech.test.azure.dfd.Main.main(Main.java:36)
Caused by: com.microsoft.azure.sdk.iot.device.exceptions.BadFormatException: Sas URI response message had no payload
	at com.microsoft.azure.sdk.iot.device.IotHubStatusCode.getConnectionStatusException(IotHubStatusCode.java:40)
	... 3 more

```

**Comment by timtay-microsoft** [comment]
Sure, it seems reasonable to nest an IotHubServiceException within the IOException in this context. I'll do the same for the completeFileUpload function as well since it seems to have the same issue of ignoring the service status code for non 200 level codes

**Comment by sanjay-90210** [comment]
@timtay-microsoft - Thank you

**Comment by timtay-microsoft** [comment]
Closing this issue as the fix has been released as of device client version 1.30.0
